### PR TITLE
cgen: fix assignment for sumtypes with pointers #6

### LINKED ISSUE #5306: c error on sum type function arguments passed by reference
**V version:** 0.1.27 3664bea
**OS:** Linux 5.6.15

**What did you do?**
```v
struct CellStr { str string }

struct CellInt { itg i64 }

struct CellFloat { flt f64 }

type Cell = CellStr | CellInt | CellFloat

fn print_cell(cell &Cell) {
	match cell {
		CellStr { println(cell.str) }
		CellInt { println(cell.itg) }
		CellFloat { println(cell.flt) }
	}
}

fn main() {
    str_cell := &CellStr{ str: 'oi' }
    itg_cell := &CellInt{ itg: 1234567890 }
    flt_cell := &CellFloat{ flt: 3.141592653 }
    print_cell(str_cell)
    print_cell(itg_cell)
    print_cell(flt_cell)
}
```

**What did you expect to see?**
```
oi
1234567890
3.1415992653
 ```
**What did you see instead?**
```
/home/runner/.cache/v/main.tmp.c: In function 'print_cell':
/home/runner/.cache/v/main.tmp.c:7650:10: error: 'cell' is a pointer; did you mean to use '->'?
  if (cell.typ == 53 /* CellStr */) {
          ^
          ->
/home/runner/.cache/v/main.tmp.c:7651:31: error: 'cell' is a pointer; did you mean to use '->'?
   CellStr* it = (CellStr*)cell.obj; // ST it
                               ^
                               ->
/home/runner/.cache/v/main.tmp.c:7654:16: error: 'cell' is a pointer; did you mean to use '->'?
  }else if (cell.typ == 54 /* CellInt */) {
                ^
...
==================
(Use `v -cg` to print the entire error message)
```

### PR DESCRIPTION
**Pull request by kalapalo**
This is my 6th attempt :-) to try to fix assignments for sum types, when either the sum type or the variant is a pointer. While JoeC is still cooking his new struct & sum type implementation, I will keep trying ;-)
The reason I want to fix this so badly is because I need to use it in a piece of code where pointers with sum types are crucial;

Cases:
The way I see, it is necessary to treat 4 different cases:
   1. sum type is pointer & variant is pointer
   2. sum type is pointer & variant is not
   3. sum type is not and variant is pointer
   4. sum type is not and variant is not
Presently, V covers #3 and #4. This fix intends to extend to #1 and #2.

Difficulties:
1. assignment needs to be done in a single statement. this is an assumption (fair one) in the several places where it is called (match, array, etc.);
2. because of #1, for the case the sum type is a pointer previously allocated, I just cannot use the current allocation, by doing 2 assignments in C (`st->obj = something;` and then `st->typ = something_else;`);
3. the only solution for #2 (sum type is pointer), and being compliant with #1, is to do a new allocation, like in the initialization (with memdup). that is how this fix is doing;
4. the drawback to this single statement approach is that there is no way to free the previous allocation, neither to the sum type nor to the obj;

Implementation:
1. there is NO shallow copy of variant pointers. the sum type .obj just points directly to it
2. when variants are not pointers, a copy of it is made on heap (memdup) and the sum type .obj points to it (V already is doing this);

this fixes #5306 and for the first time passes all my tests (assignments to and from sum types, arrays of sum types, recursive fn, and pointers to pointers)

### CODE REVIEW

**Comment by joe-conigliaro** [comment]
Hi @kalapalo when you say `The reason I want to fix this so badly is because I need to use it in a piece of code where pointers with sum types are crucial` can you please provide sample code to illustrate the problem.

Also when you say `variant is pointer` do you mean the sum type definitions is as follows?:
```
type MySumType = int | string | &Foo
```
which would mean that variant must be a pointer to type Foo

Thanks @kalapalo

**Comment by joe-conigliaro** [comment]
@kalapalo Can you please provide some sample code or test illustrating the issue this solves. Thanks

**Comment by kalapalo** [comment]
Hey Joe. Thanks for checking this.
I have not though of `type MySumType = int | string | &Foo` and in my big ignorance, I don't see a need for this since you can do `type MySumType = int | string | Foo` and the sum type could receive pointers and not pointers.
Maybe when you do `type MyOtherSumType = &Foo| Bar | Baz` you are saying only `&Foo` are allowed and `Foo` are not. If this is the case, then `&Bar` and `&Baz` would also not be allowed. Hmm, I don't see this as the direction V is pointing ... (maybe I am wrong, what would not be a surprise at all :-)) )
File sent through discord.

**Comment by kalapalo** [comment]
file:///tmp/test_sumtype_ptr.v

**Comment by kalapalo** [comment]
here it is, just in case:

[test_sumtype_ptr.v.zip](https://github.com/vlang/v/files/4949168/test_sumtype_ptr.v.zip)

**Comment by joe-conigliaro** [comment]
@kalapalo I was asking in case thats what you meant, at the time I wasn't sure. Thanks for the reply

**Comment by kalapalo** [comment]
@joe-conigliaro , I was thinking about your comments above, about having sum types like `type MySumType = str | u32 | bool`. I have to confess that I really was not thinking about primitive types, my bad, and thank you for take my attention to it. This fix is not covering that, this is a fact.
But I think with primitive types, things start getting interesting, because this touches the foundations of V.
For structs, we have dereferencing on the fly, by that I mean that I can pass a `Abc` struct to a fn with a by ref type arg &Abc and this is no problema, but with primitive types this is not allowed, right ?
Basically, for this fix to cover primitive types, according to the original goal (covering pointers and not pointers), I would have to also deref primitive type pointers and I would compare things like &my_str with "hello", what, as I understand, breaks some V rules.
Maybe I am complicating what should not be. Please give me your thoughts on that.
Thanks.
